### ISSUE TITLE
Init Container not translated from k8s platform to task's pod

### ISSUE THREAD

**Issue by juanpablo-santos** [spring-cloud/spring-cloud-dataflow#5280]
**Description:**
Init containers defined at k8s platform level don't get propagated to applications running on that platform, they must be set via task properties

**Release versions:**
```json
{
  "versions": {
    "implementation": {
      "name": "spring-cloud-dataflow-server",
      "version": "2.10.2"
    },
    "core": {
      "name": "Spring Cloud Data Flow Core",
      "version": "2.10.2"
    },
    "dashboard": {
      "name": "Spring Cloud Dataflow UI",
      "version": "3.3.2"
    },
    "shell": {
      "name": "Spring Cloud Data Flow Shell",
      "version": "2.10.2",
      "url": "https://repo.maven.apache.org/maven2/org/springframework/cloud/spring-cloud-dataflow-shell/2.10.2/spring-cloud-dataflow-shell-2.10.2.jar"
    }
  },
  "features": {
    "streams": true,
    "tasks": true,
    "schedules": true,
    "monitoringDashboardType": "NONE"
  },
  "runtimeEnvironment": {
    "appDeployer": {
      "deployerImplementationVersion": "2.9.1",
      "deployerName": "Spring Cloud Skipper Server",
      "deployerSpiVersion": "2.9.2",
      "javaVersion": "1.8.0_362",
      "platformApiVersion": "",
      "platformClientVersion": "",
      "platformHostVersion": "",
      "platformSpecificInfo": {
        "default": "kubernetes"
      },
      "platformType": "Skipper Managed",
      "springBootVersion": "2.7.9",
      "springVersion": "5.3.25"
    },
    "taskLaunchers": [
      {
        "deployerImplementationVersion": "2.8.2",
        "deployerName": "KubernetesTaskLauncher",
        "deployerSpiVersion": "2.8.2",
        "javaVersion": "1.8.0_362",
        "platformApiVersion": "v1",
        "platformClientVersion": "unknown",
        "platformHostVersion": "unknown",
        "platformSpecificInfo": {
          "namespace": "scdf",
          "master-url": "https://XXXYYYZZZ:6443"
        },
        "platformType": "Kubernetes",
        "springBootVersion": "2.7.9",
        "springVersion": "5.3.25"
      },
      {
        "deployerImplementationVersion": "2.8.2",
        "deployerName": "KubernetesTaskLauncher",
        "deployerSpiVersion": "2.8.2",
        "javaVersion": "1.8.0_362",
        "platformApiVersion": "v1",
        "platformClientVersion": "unknown",
        "platformHostVersion": "unknown",
        "platformSpecificInfo": {
          "namespace": "scdf",
          "master-url": "https://XXXYYYZZZ:443/"
        },
        "platformType": "Kubernetes",
        "springBootVersion": "2.7.9",
        "springVersion": "5.3.25"
      }
    ]
  },
  "monitoringDashboardInfo": {
    "url": "",
    "source": "default-scdf-source",
    "refreshInterval": 15
  },
  "security": {
    "isAuthentication": false,
    "isAuthenticated": false,
    "username": null,
    "roles": []
  }
}
```

**Custom apps:**
N/A

**Steps to reproduce:**
Define a platform that defines init-containers, such as:
```yaml
spring:
  cloud:
    dataflow:
      task:
        platform:
          kubernetes:
            accounts:
              onp:
                fabric8:
                  clientCertData: "client-cert-data-here"
                  clientKeyData: "client-key-data-here"
                  masterUrl: "https://XXX.YYY:6443"
                  namespace: "scdf"
                  trustCerts: true
                imagePullPolicy: Always
                initContainer:
                  containerName: shared-dir
                  environmentVariables: SHARED_DIR=/shared-dir
                  commands: /tmp/init.sh
                  imageName: XXXX.YYYY:zzz
                  volumeMounts:
                    - name: vol-shared-dir
                      mountPath: "/shared-dir"
                      readOnly: false
                namespace: "scdf"
                volumes:
                  - name: "vol-shared-dir"
                volumeMounts:
                  - name: "vol-shared-dir"
                    mountPath: "/shared-dir"
                    readOnly: false
```

When running any task on that platform, you'll see that, by default, every parameter (env vars, volumes, volume mounts, pull policy, fabric8 connection details, etc) in the above configuration is built into the pod, except the init-containers. We have to manually pass them as task properties, something along the lines of:
```
spring.cloud.dataflow.task.platformName=onp
deployer.<TASK_ID>.kubernetes.init-container.container-name=shared-dir
deployer.<TASK_ID>.kubernetes.init-container.environment-variables=SHARED_DIR=/shared-dir
deployer.<TASK_ID>.kubernetes.init-container.commands=/tmp/init.sh
deployer.<TASK_ID>.kubernetes.init-container.image-name=XXXX.YYYY:zzz
deployer.<TASK_ID>.kubernetes.init-container.volume-mounts=[{name: 'vol-shared-dir', mountPath: '/shared-dir', readOnly: false}]
```

We've also tried setting the properties using the `spring.cloud.deployer.kubernetes` namespace, and IIRC, `deployer.kubernetes`, but that didn't make a difference.

This happens setting up the server either as a local standalone jar, or through the helm chart.

**Screenshots:**
N/A

**Additional context:**
N/A

**Comment by corneil** [comment]
The initContainer property must be populated using a JSON representation like:
```json
{
  "containerName": "shared-dir",
  "env": [
    {
      "name": "SHARED_DIR",
      "value": "/shared-dir"
    }
  ],
  "command": [
    "sh",
    "/tmp/init.sh"
  ],
  "image": "XXXX.YYYY:zzz",
  "volumeMounts": [
    {
      "name": "vol-shared-dir",
      "mountPath": "/shared-dir",
      "readOnly": false
    }
  ]
}
```

Thus the following should give you what you need:

```properties
deployer.<TASK_ID>.kubernetes.initContainer={"containerName":"shared-dir","env":[{"name":"SHARED_DIR","value":"/shared-dir"}],"command":["sh","/tmp/init.sh"],"image":"XXXX.YYYY:zzz","volumeMounts":[{"name":"vol-shared-dir","mountPath":"/shared-dir","readOnly":false}]}
```

**Comment by juanpablo-santos** [comment]
Hi @corneil , 

I had to set `env` as `environmentVariables`, and its values as a list of `NAME=VALUE`, instead of as a map, and `command` was `commands` but yes, indeed setting the initContainer as json did the trick, now it's being applied by default :-) 

thanks for your insights and prompt response!

**Comment by juanpablo-santos** [comment]
Hi, 

so after setting the initContainer as json on the platform, seems that this bit of deployment isn't translated to the target platform on first run, and we have to set it manually as deployer properties for our tasks, but only once. After doing that, the next executions of the same task do pass the initContainer to the target platform, so we don't have to pass it as deployer properties again. 

Is this information cached somewhere, or stored in the db? It happens for every new task that we register on our SCDF instance.

so just to summarize:
* we defined a k8s platform with unvalid initContainer info -> we had to pass this as deployer properties
* we fixed the initContainer bit and redeployed our SCDF server -> we had to pass the initContainer info as deployer properties on first execution, afterwards is not needed, so we suspect this is info is cached / stored / versioned somewhere?

**Comment by cppwfs** [comment]
Hello @juanpablo-santos ,
When you are using properties (deployer or application) for a task definition, each subsequent launch of that task definition will reuse the properties from the previous execution. This is a part of the CI/CD strategy for Tasks.

**Comment by juanpablo-santos** [comment]
Hi @cppwfs ,

yes, but then why on first run, initContainer isn't getting passed? We had to different cases
* we defined a k8s platform with unvalid initContainer info -> we had to pass this as deployer properties _each_ time
* we fixed the initContainer bit and redeployed our SCDF server -> we have to pass this only on _first_ time. Subsequent runs reuse the properties, and that works as expected, but on first run, the initContainer isn't getting passed

best regards, and thanks for your prompt response!

**Comment by cppwfs** [comment]
There is an existing story where a user can create a task definition with both app and deployer properties.   
https://github.com/spring-cloud/spring-cloud-dataflow/issues/3962

**Comment by onobc** [comment]
This is actually a bug on the k8s deployer where it is not taking the account properties into consideration when it constructs the init container. 

https://github.com/spring-cloud/spring-cloud-deployer-kubernetes/blob/529ba45d5a3ccfe35489cb41c53ce07f216e9f37/src/main/java/org/springframework/cloud/deployer/spi/kubernetes/DeploymentPropertiesResolver.java#L591 
should also check if `this.properties.getInitiContainer() != null` as `this.properties` holds the global account properties. They are currently ignored.
